### PR TITLE
feat: JSX quotes option

### LINKED ISSUE #1080: Provide way to use single quotes in jsx
I read [this comment](https://github.com/prettier/prettier/issues/73#issuecomment-271953751) saying that "I've never seen anyone write `<div className='a' />`". Well, here's someone 👋 

The same way `trailingComma` accepts a string, we could let `singleQuote` do as well. 
- `'all'` would use single quotes in js as well as jsx
- `'js'` (or `true`) would use single quotes in js but not in jsx
- `'jsx'` would use single quotes in jsx but not in js

### PR DESCRIPTION
**Pull request by baelec**
Fixes #1080.

This pull request adds support for customizing quotes in JSX. While this has been discussed before, I feel this is worth revisiting.

It was mentioned in [the past](https://github.com/prettier/prettier/issues/656#issuecomment-306321171) that if people wanted to have this flexibility, they'd have to fork Prettier. Since this is a pretty common feature request that helps simplify people's lives and it only adds a trivial amount of code to maintain, my hope is that there can be a discussion about merging this into Prettier.

Thoughts? Suggestions?

### CODE REVIEW

**Comment by baelec** [comment]
Also addresses [this open issue.](https://github.com/prettier/prettier/issues/1080)

File: package-lock.json [hunk 1]
```diff
@@ -0,0 +1,8213 @@
+{
```
Comment (lipis) [inline package-lock.json]: Remove the `package-lock.json` as we are using Yarn.. :)

File: src/language-js/options.js [hunk 2]
```diff
@@ -52,6 +52,13 @@ module.exports = {
     default: false,
     description: "Use single quotes instead of double quotes."
   },
+  jsxSingleQuote: {
+    since: "1.10.3",
+    category: CATEGORY_JAVASCRIPT,
+    type: "boolean",
+    default: false,
+    description: "Use single quotes instead of double quotes."
```
Comment (j-f1) [inline src/language-js/options.js]: > Use single quotes instead of double quotes **in JSX**

File: src/language-js/printer-estree.js [hunk 3]
```diff
@@ -1631,12 +1631,19 @@ function printPathNoParens(path, options, print, args) {
...
+          let quote = '"';
+          let escapedChar = "&quot;";
+          if (options.jsxSingleQuote) {
+            regex = /'/g;
+            escapedChar = "&apos;";
+            quote = "'";
+          }
           const value = rawText(n.value);
-          res = '"' + value.slice(1, -1).replace(/"/g, "&quot;") + '"';
+          res = quote + value.slice(1, -1).replace(regex, escapedChar) + quote;
```
Comment (j-f1) [inline src/language-js/printer-estree.js]: Can you add something here to replace the `&apos;` in `'foo &apos; bar'` with `'` when `jsxSingleQuote` is false?

File: tests/jsx_quotes/jsfmt.spec.js [hunk 4]
```diff
@@ -0,0 +1,2 @@
+run_spec(__dirname, ["flow", "typescript"], { jsxSingleQuote: true });
+run_spec(__dirname, ["flow", "typescript"], { jsxSingleQuote: false });
```
Comment (j-f1) [inline tests/jsx_quotes/jsfmt.spec.js]: Please add `"babylon"` to both of these tests — it’s actually a different parser from Flow’s parser.

File: tests/jsx_quotes/test.js [hunk 5]
```diff
@@ -0,0 +1,6 @@
+<SimpleComponent
+  str0="abc"
+  str1='abc'
+  str2="a'b'c"
+  str3='a"b"c'
```
Comment (j-f1) [inline tests/jsx_quotes/test.js]:
```jsx
  str4="a&quot;b&quot;c"
  str5='a&apos;b&apos;c'
```

File: src/language-js/printer-estree.js (hunk 3 above)
Comment (baelec) [reply src/language-js/printer-estree.js]: Hey j-f1, I can do that. In a similar manner, would we also want to replace `&quot;` in  `"foo &quot; bar"` with `"` when `jsxSingleQuote` is true?

Comment (j-f1) [reply src/language-js/printer-estree.js]: That would be great!

File: tests/jsx_quotes/test.js (hunk 5 above)
Comment (baelec) [reply tests/jsx_quotes/test.js]: Not sure why Github is still showing this as outstanding but the change has been made.

**Comment by baelec** [comment]
All requested changes have been made. Thanks for the feedback! Let me know if there's anything else that needs to be done.

File: tests/jsx_quotes/test.js (hunk 5 above)
Comment (lipis) [reply tests/jsx_quotes/test.js]: Because the comment was made above the changes.. :)

Comment (baelec) [reply tests/jsx_quotes/test.js]: Yeah, I would have just expected it to go away after the changes were made. See: [here](https://github.com/baelec/prettier/blob/428d0e772cb01d344595c92c5768a9843520b640/tests/jsx_quotes/test.js).

Comment (j-f1) [reply tests/jsx_quotes/test.js]: GitHub only collapses comments when the line the comment is on is changed.

Don’t worry about it too much, it’s not a problem :)

**Comment by maxharris9** [comment]
Has someone forked Prettier yet? There are a great number of us that need it to work with Standard and JSX. We could make it do the right thing with semicolons, single quotes, etc. with no configuration!

Let's do it!

**Comment by j-f1** [comment]
@maxharris9 Check out [`prettier-miscellaneous`](https://github.com/arijs/prettier-miscellaneous).

**Comment by suchipi** [comment]
@maxharris9 there is https://github.com/arijs/prettier-miscellaneous, but it is unmaintained.

**Comment by suchipi** [comment]
I'm going to close this since we're not going to merge it.

**Comment by maxharris9** [comment]
@j-f1 thanks! that looks just like what I need!

**Pull request closed without merging**